### PR TITLE
Standard library imports style check

### LINKED ISSUE #49647: STYLE place standard library imports at top of file
Imports should typically be placed at the top of files. Sometimes, imports are placed inside functions to:
- avoid circular imports
- avoid `ImportError` if it's an optional dependency

Standard library imports should really always be at the top of files.

Noticed in https://github.com/pandas-dev/pandas/pull/49645 that this is often not the case

I've made a script to automate detecting when this is the case. So the task is:
```
git checkout -b standard-library-imports main
git pull git@github.com:MarcoGorelli/pandas.git standard-library-imports
git reset --hard FETCH_HEAD
pre-commit run stdlib-imports --all-files
```
Then, fixup any errors that are reported. Finally, stage your changes, commit them, push them to your fork, and open a pull request

Feel free to reach out if you into any issues along the way

If any wants to take this, it would be a nice and welcome clean up!

---

EDIT: after going through a PR, I'm not sure it's worth introducing a check for this - but we can still take some of the cleanups it found

### PR DESCRIPTION
**Pull request by Thextan**
- [X] closes #49647
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

### CODE REVIEW

**Comment by MarcoGorelli** [comment]
you can push to the same branch, no need to open a new pull request

**Comment by Thextan** [comment]
@MarcoGorelli I closed the previous pull request and merged the two sets of changes into this PR.

**Comment by MarcoGorelli** [comment]
OK - in the future a single PR is fine, you can just add extra commits to it

**Comment by MarcoGorelli** [comment]
there's a few pylint failures: https://github.com/pandas-dev/pandas/actions/runs/3791678410/jobs/6447285304

I'm thinking it might not be worth introducing this lint check, sorry - if you want to clean up the import changes in the tests (such that pylint still passes), then we could take that

**Comment by Thextan** [comment]
@MarcoGorelli please look at this PR.  I have two code tests failing that I can't seem to fix without making it worse.  Code Checks/Docstring: I am not sure how to get around the duplication between xml.etree... (line 13) and lxml.etree... (lines 517/590).

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
I don't think the check is worth adding anymore, sorry (I hadn't realised at the start how expensive some of them were, nor that it would conflict with the more important pylint check) - but the cleanups in the test folder are nice

How about
```
git fetch upstream
git merge upstream/main
git reset upstream/main
git add pandas/tests
git checkout -- .
git commit -a -m 'put stdlib imports in pandas/tests at top of file'
git push origin HEAD -f
```
and then we can get that in?

**Comment by Thextan** [comment]
Requested changes have been made and pushed.

File: pandas/tests/frame/test_constructors.py [hunk 1]
```diff
@@ -1,13 +1,18 @@
+import array
 from collections import (
     OrderedDict,
     abc,
+    defaultdict,
 )
+from collections import namedtuple  # GH11181
```
Comment (MarcoGorelli) [inline pandas/tests/frame/test_constructors.py:7]: this comment should stay with the test

File: pandas/tests/groupby/test_timegrouper.py [hunk 2]
```diff
@@ -1,6 +1,10 @@
-""" test with the TimeGrouper / grouping with datetimes """
-
-from datetime import datetime
+"""
+test with the TimeGrouper / grouping with datetimes
+"""
+from datetime import (
+    datetime,
+    timedelta as dttd,
```
Comment (MarcoGorelli) [inline pandas/tests/groupby/test_timegrouper.py:6]: does it work to just import `timedelta`, without an alias?

File: pandas/tests/io/pytables/test_store.py [hunk 3]
```diff
@@ -1,6 +1,8 @@
-import datetime
+import datetime as dt
+from datetime import datetime
```
Comment (MarcoGorelli) [inline pandas/tests/io/pytables/test_store.py:2]: there shouldn't be any need for both, just use `dt.datetime`

File: pandas/tests/io/test_orc.py [hunk 4]
```diff
@@ -1,5 +1,8 @@
-""" test orc compat """
+"""
+test orc compat
+"""
```
Comment (MarcoGorelli) [inline pandas/tests/io/test_orc.py:3]: could you revert unrelated changes (such as this one) please?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
getting there 💪

File: pandas/tests/io/test_orc.py (hunk 4 above)
Comment (Thextan) [reply pandas/tests/io/test_orc.py:3]: git reset on this file.

File: pandas/tests/io/pytables/test_store.py (hunk 3 above)
Comment (Thextan) [reply pandas/tests/io/pytables/test_store.py]: Removed second import and used dt.datetime.  I renamed variable dt to mydt.

File: pandas/tests/groupby/test_timegrouper.py (hunk 2 above)
Comment (Thextan) [reply pandas/tests/groupby/test_timegrouper.py]: Used timedelta without  the alias.

File: pandas/tests/frame/test_constructors.py (hunk 1 above)
Comment (Thextan) [reply pandas/tests/frame/test_constructors.py]: Moved comment back to test.

**Comment by MarcoGorelli** [comment]
should be good, thanks, just pending green

**Review by MarcoGorelli — APPROVED** [review]
thanks @Thextan
